### DISCUSSION TITLE
How to handle custom methods in endpoint routes such as /posts/{id}:like and /posts/{id}:unlike

### DISCUSSION THREAD

**Discussion by erenz14** [fastapi/fastapi#6102] [Questions] [answered]
### First Check

- [X] I added a very descriptive title to this issue.
- [X] I used the GitHub search to find a similar issue and didn't find it.
- [X] I searched the FastAPI documentation, with the integrated search.
- [X] I already searched in Google "How to X in FastAPI" and didn't find any information.
- [X] I already read and followed all the tutorial in the docs and didn't find an answer.
- [X] I already checked if it is not related to FastAPI but to [Pydantic](https://github.com/samuelcolvin/pydantic).
- [X] I already checked if it is not related to FastAPI but to [Swagger UI](https://github.com/swagger-api/swagger-ui).
- [X] I already checked if it is not related to FastAPI but to [ReDoc](https://github.com/Redocly/redoc).

### Commit to Help

- [X] I commit to help with one of those options 👆

### Example Code

```python
from typing import Union

from fastapi import FastAPI

app = FastAPI()

@app.patch("/posts/{id}:like")
def like_post(id: int):
    return {"id": id, "message": "This post was successfully liked!"}

@app.patch("/posts/{id}:unlike")
def unlike_post(id: int):
    return {"id": id, "message": "This post was successfully unliked!"}
```

### Description

What is the best way to define a custom method that falls outside of the standard HTTP verbs? For example liking or unliking a post on a social media website.

I tried the included code sample and I get an error: "value is not a valid integer" because I think it is trying to validate the {id}:unlike as an integer.

Google provides some details on custom methods here and I'm trying to use them in the way they describe: https://cloud.google.com/apis/design/custom_methods

### Operating System

Windows

### FastAPI Version

0.78.0

### Python Version

3.8.0

**Comment by jgould22** [comment]
Don't over think it.

The resource is a `like`, dont patch the post, create or delete a like.

```
# Create a like for a specific post
@app.post("/posts/{id}/like

# Delete (unlike) a like for a specific post
@app.delete("/posts/{id}/like
```

**Comment by erenz14** [comment]
Thanks for the reply!

Yes that or something similar would work for this specific example. But a couple reasons I think it'd still be helpful to know the best way to handle custom methods in the way defined by Google:

1. Some companies might mandate a standard for custom methods in the way Google defined it
2. Maybe there are some use cases out there where custom methods are more suitable than the standard methods

Obviously if there is no way to implement custom methods cleanly then that is fine too :)

**Comment by jgould22** [comment]
Ah yes.

So the behavior you are seeing is described here in starlette https://www.starlette.io/routing/#path-parameters 

You may be able to achieve what you want with a custom converter (I am not sure this is exposed in FastAPI) and a custom router but I don't think it will work out of the box.

**Comment by b0g3r** [comment]
@erenz14 could you share curl code which are returning "value is not a valid integer" error? Because I don't get this error:

```python
@app.get("/api/{id:int}:action1")
async def action1(id: int) -> int:
    return id

@app.get("/api/{id:int}:action2")
async def action2(id: int) -> int:
    return id
```

```bash
> curl -X 'GET' \
  'http://127.0.0.1:8000/api/100:action1' \
  -H 'accept: application/json'
100
```

**Comment by erenz14** [comment]
> @erenz14 could you share curl code which are returning "value is not a valid integer" error? Because I don't get this error:
> 
> ```python
> @app.get("/api/{id:int}:action1")
> async def action1(id: int) -> int:
>     return id
> 
> @app.get("/api/{id:int}:action2")
> async def action2(id: int) -> int:
>     return id
> ```
> 
> ```shell
> > curl -X 'GET' \
>   'http://127.0.0.1:8000/api/100:action1' \
>   -H 'accept: application/json'
> 100
> ```

### This is the curl I've been using:

```
$ curl --location --request PATCH 'http://localhost:8000/posts/1:like'
  % Total    % Received % Xferd  Average Speed   Time    Time     Time  Current
                                 Dload  Upload   Total   Spent    Left  Speed
100    99  100    99    0     0  18907      0 --:--:-- --:--:-- --:--:-- 33000{"detail":[{"loc":["path","id"],"msg":"value is not a valid integer","type":"type_error.integer"}]}
```

### Using your example I noticed the following behavior. None of them actually returned the ID.

**Using your exact example that had the route as "/api/{id:int}:action1":**
 
```
$ curl -X 'GET'   'http://127.0.0.1:8000/api/100:action1'   -H 'accept: application/json'
  % Total    % Received % Xferd  Average Speed   Time    Time     Time  Current
                                 Dload  Upload   Total   Spent    Left  Speed
100    22  100    22    0     0   3760      0 --:--:-- --:--:-- --:--:--  5500{"detail":"Not Found"}
```
 
**Using your example but removing ":int" from route i.e. "/api/{id}:action1":**

```
$ curl -X 'GET'   'http://127.0.0.1:8000/api/100:action1'   -H 'accept: application/json'
  % Total    % Received % Xferd  Average Speed   Time    Time     Time  Current
                                 Dload  Upload   Total   Spent    Left  Speed
100    99  100    99    0     0  13996      0 --:--:-- --:--:-- --:--:-- 19800{"detail":[{"loc":["path","id"],"msg":"value is not a valid integer","type":"type_error.integer"}]}
```
@jgould22 Thanks for the links. I'm going to spend some time today exploring what you referenced and will close this issue out depending on what I discover.

**Comment by YuriiMotov** [comment, marked as answer]
Works well in current version:

```python
from fastapi import FastAPI

app = FastAPI()

@app.get("/api/{id:int}:action1")
async def action1(id: int):
    return {"action1": id}

@app.get("/api/{id:int}:action2")
async def action2(id: int):
    return {"action2": id}

# curl -X 'GET' 'http://localhost:8000/api/777:action1' -H 'accept: application/json'
# {"action1":777}

# curl -X 'GET' 'http://localhost:8000/api/777:action2' -H 'accept: application/json'
# {"action2":777}
```
